### PR TITLE
Using composition instead of inheritance

### PR DESCRIPTION
**Pull request by cantrowitz**

### CODE REVIEW

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/GlobalWPermissionsBroadcastProviderRegistration.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.IntentFilter;
+import android.os.Handler;
+
+/**
+ * Created by adamcantrowitz on 9/1/15.
+ */
+class GlobalWPermissionsBroadcastProviderRegistration implements BroadcastReceiverRegistrationStrategy {
```
Comment (JakeWharton) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/GlobalWPermissionsBroadcastProviderRegistration.java]: What's with the "W"?

Comment (cantrowitz) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/GlobalWPermissionsBroadcastProviderRegistration.java]: 'With"

**Comment by cantrowitz** [comment]
@JakeWharton @DLew, any more issues, or is this good?

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java [hunk 2]
```diff
@@ -16,12 +15,12 @@
 /**
  * Created by adamcantrowitz on 9/1/15.
  */
- abstract class BroadcastProviderStrategy implements Observable.OnSubscribe<Intent> {
-    protected final IntentFilter intentFilter;
+class BroadcastProvider implements Observable.OnSubscribe<Intent> {
     private BroadcastReceiver broadcastReceiver;
+    private BroadcastRegistrarStrategy broadcastRegistrarStrategy;
```
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java]: I'd make these `final` if you can.

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java [hunk 3]
```diff
@@ -7,22 +7,23 @@
 /**
  * Created by adamcantrowitz on 9/1/15.
  */
-class GlobalBroadcastProvider extends BroadcastProviderStrategy {
+class BroadcastRegistrar implements BroadcastRegistrarStrategy {
 
     protected final Context context;
```
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java]: Nit: Why is this `protected`?

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java [hunk 4]
```diff
@@ -7,22 +7,23 @@
...
+    private final IntentFilter intentFilter;
 
-    GlobalBroadcastProvider(IntentFilter intentFilter, Context context) {
-        super(intentFilter);
+    BroadcastRegistrar(IntentFilter intentFilter, Context context) {
```
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java]: Supernit: Params could match order above.

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java [hunk 5]
```diff
@@ -32,29 +31,28 @@ public void call(final Subscriber<? super Intent> subscriber) {
...
+    BroadcastReceiver getBroadcastReceiver() {
         return broadcastReceiver;
     }
 
-    protected abstract void registerBroadcastReceiver(BroadcastReceiver broadcastReceiver);
-
-    protected abstract void unregisterBroadcastReceiver(BroadcastReceiver broadcastReceiver);
 
     @NonNull
     private BroadcastReceiver createBroadcastReceiver(final Subscriber<? super Intent> subscriber) {
```
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java:53]: Do you need this method now? Could just create it inline...

**Comment by dlew** [comment]
I know you had doubts but this feels a lot better to me. Each registrar is focused on its job and that makes it easy to read through.

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java (hunk 5 above)
Comment (cantrowitz) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java:53]: I try and have my methods do one thing. I find that's it's easier to read and maintain. I think this is largely personal and has no impact on the function of this area. 

"One way to know that a function is doing more than 'one thing' is if you can extract another function from it with a name that is not merely a restatement of its implementation." -- Clean Code by Robert C. Martin

If it is more of the standard of RX to do it inline, I'd be happy to do that as well.

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java (hunk 4 above)
Comment (cantrowitz) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java]: Sorry, I'm not following. Can you elaborate?

**Comment by cantrowitz** [comment]
Thanks for the comments @dlew!

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java (hunk 5 above)
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastProvider.java:53]: I'm down with that. I just figured the reason for it was the previous subclassing but I guess not.

File: rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java (hunk 4 above)
Comment (dlew) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java]: Oh, this is just super nit-picky, but the params above are in the order Context, IntentFilter but the constructor goes in the opposite order.

This is some serious bike shedding so feel free to ignore.

Comment (cantrowitz) [inline rxbroadcast/src/main/java/com/cantrowitz/rxbroadcast/BroadcastRegistrar.java]: Ah, I see now. I agree. Will fix.

**Pull request merged**